pypsa/plot.py

```diff
@@ -430,151 +345,507 @@ def plot(
+
+    # Add missing colors
+    # TODO: This is not consistent, since for multiindex a ValueError is raised
+    if not multindex_buses:
+        bus_colors = bus_colors.reindex(n.buses.index)
+
+    # Raise additional ValueErrors after formatting
+    if multindex_buses:
+        if len(bus_sizes.index.unique(level=0).difference(n.buses.index)) != 0:
+            msg = "The first MultiIndex level of sizes must contain buses"
+            raise ValueError(msg)
+        if not bus_sizes.index.unique(level=1).isin(bus_colors.index).all():
+            msg = "Colors not defined for all elements in the second MultiIndex "
+            "level of sizes, please make sure that all the elements are "
+            "included in colors or in n.carriers.color"
+            raise ValueError(msg)
+
+    # Apply all cmaps
+    bus_colors = _apply_cmap(bus_colors, bus_cmap, bus_cmap_norm)
+    line_colors = _apply_cmap(line_colors, line_cmap, line_cmap_norm)
+    link_colors = _apply_cmap(link_colors, link_cmap, link_cmap_norm)
+    transformer_colors = _apply_cmap(
+        transformer_colors, transformer_cmap, transformer_cmap_norm
     )
 
+    # Initiate NetworkPlotter
+    plotter = NetworkPlotter(n)
+    plotter.init_layout(layouter)
+    buses = bus_sizes.index if not multindex_buses else bus_sizes.index.unique(level=0)
+    plotter.init_boundaries(buses, boundaries, margin)
+    plotter.init_axis(ax, projection, geomap, geomap_colors, title)
+
+    # Add jitter if given
+    if jitter is not None:
+        plotter.add_jitter(jitter)
+
+    # Plot buses
+    bus_sizes = bus_sizes.sort_index(level=0, sort_remaining=False)
+    if geomap:
+        bus_sizes = bus_sizes * plotter.area_factor**2
+    if isinstance(bus_sizes.index, pd.MultiIndex):
+        patches = plotter.get_multiindex_busses(
+            bus_sizes, bus_colors, bus_alpha, bus_split_circles
+        )
+    else:
+        patches = plotter.get_singleindex_busses(bus_sizes, bus_colors, bus_alpha)
+    bus_collection = PatchCollection(patches, match_original=True, zorder=5)
+    plotter.ax.add_collection(bus_collection)
+
+    # Collect data for branches and flows
+    branch_data = {
+        "Line": {
+            "width": line_widths,
+            "color": line_colors,
+            "alpha": line_alpha,
+        },
+        "Link": {
+            "width": link_widths,
+            "color": link_colors,
+            "alpha": link_alpha,
+        },
+        "Transformer": {
+            "width": transformer_widths,
+            "color": transformer_colors,
+            "alpha": transformer_alpha,
+        },
+    }
+    if branch_components is None:
+        branch_components = plotter.n.branch_components
+
+    # Plot branches
+    branches = plotter.get_branch_collections(
+        branch_components,
+        geometry,
+        branch_data,
+    )
+    for branch in branches.values():
+        plotter.ax.add_collection(branch)
+
+    # Plot flows
```

Comment:
this still feels a bit weird that we are defining all branch components separately to then put them together into the get_branch_collections function where we iterate over them again. I would vote for making get_branch_collection process one branch component at a time